### PR TITLE
extensions: Add extensions metadata.yaml

### PR DESCRIPTION
**Pull request by phlax**
Signed-off-by: Ryan Northey <ryan@synca.io>

Commit Message: extensions: Add extensions metadata.yaml
Additional Description:

This is a breakout pr from #16496 as that will require a lot of change

This PR:

- adds an `extensions_metdata.yaml` file
- uses the new metadata file in `protodoc`
- removes `generate_extension_db.py`
- adds `validate_extension_db.py` to
    - ensure that old <> new metadata match
    - consolidate existing metadata validation in one place

Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by phlax** [comment]
@htuch this is hopefully ready for review

File: docs/requirements.txt
Comment (dmitri-d) [inline docs/requirements.txt:38]: Nit: if you could make changes to `docs/requirements.txt` in a separate PR, it would make life slightly easier for maintainers downstream.

**Comment by dmitri-d** [comment]
Could you mention how to update metadata in https://github.com/envoyproxy/envoy/blob/main/CONTRIBUTING.md#adding-new-extensions?

**Comment by dmitri-d** [comment]
looks good, other than a couple of nits above.

File: source/extensions/extensions_metadata.yaml
Comment (htuch) [inline source/extensions/extensions_metadata.yaml:574]: There were a bunch of comments around these status, i.e. that this alpha/unknown status can't change until we do deliberate review and change to threat model, for Wasm runtimes. Can you move these as well to this file? Also pleaes check we haven't missed other places there are similar comments we are missing.

File: source/extensions/extensions_metadata.yaml
Comment (htuch) [inline source/extensions/extensions_metadata.yaml:1]: I think you need to update developer docs to cover this file as well.

File: tools/extensions/validate_extensions.py
Comment (htuch) [inline tools/extensions/validate_extensions.py:6]: I think it should default to using `$PWD` and then in Docker set `/source` via `ENVOY_SRCDIR`. The only place we expect to see `/source` is in Docker.

File: tools/extensions/validate_extensions.py
Comment (htuch) [inline tools/extensions/validate_extensions.py:238]: Does this file need unit tests?

File: tools/extensions/validate_extensions.py
Comment (htuch) [inline tools/extensions/validate_extensions.py:156]: Is this just to validate we're doing the right thing as scaffolding? Should it be deleted before merging?

File: tools/extensions/validate_extensions.py
Comment (htuch) [inline tools/extensions/validate_extensions.py:231]: This is a bit weird; the exit code is some function of the number of errors? I think it should be a fixed value on failure to give some determinism.

File: source/extensions/extensions_metadata.yaml
Comment (htuch) [inline source/extensions/extensions_metadata.yaml:8]: I think I prefer to see `status` being explicitly called out as `stable` rather than defaulted; this reflects an opionionated stance; folks declaring extensions must state it upfront.

**Review by htuch — CHANGES_REQUESTED** [review]
Great, thanks, a few comments.
/wait

File: docs/requirements.txt
Comment (phlax) [reply docs/requirements.txt:38]: @dmitri-d on this point i would like to push back - adding yaml to the docs build requirements is not needed without this PR - it makes sense to add it here i think

File: source/extensions/extensions_metadata.yaml
Comment (phlax) [reply source/extensions/extensions_metadata.yaml:574]: this PR hasnt removed the other comments yet - but yep - i will go through and add any comments i can find from the existing metadata registration

File: source/extensions/extensions_metadata.yaml
Comment (phlax) [reply source/extensions/extensions_metadata.yaml:1]: yep - as @dmitri-d asked - i will add to the dev docs about updating this file

File: tools/extensions/validate_extensions.py
Comment (phlax) [reply tools/extensions/validate_extensions.py:6]: this is very temporary - in the next PR most of this will be gone

File: tools/extensions/validate_extensions.py
Comment (phlax) [reply tools/extensions/validate_extensions.py:238]: yes - can i do it in the next PR when much of this code will be removed - im more than happy to shift this to a `Checker` class and add the appropriate tests

File: tools/extensions/validate_extensions.py
Comment (phlax) [reply tools/extensions/validate_extensions.py:156]: yes and no - its to ensure that transitioning from the old macro way of setting the metadata and the new yaml way nothing is lost

File: tools/extensions/validate_extensions.py
Comment (phlax) [reply tools/extensions/validate_extensions.py:231]: yep - agreed - ill update

File: source/extensions/extensions_metadata.yaml
Comment (phlax) [reply source/extensions/extensions_metadata.yaml:8]: ok - this just reflects the default that was being set in the macro - but i agree - we shouldnt allow a default here and people should declare its status

File: tools/extensions/validate_extensions.py
Comment (phlax) [reply tools/extensions/validate_extensions.py:6]: also - re "$PWD" - this cant do that because its inside the bazel build dir - a path must be passed, there cant be a default that is not hardcoded.

this is one of the reasons for doing this - we wont need the envoy_srcdir to be set at all once this code is trimmed down - the bazel env should have anything that is required

Comment (phlax) [reply tools/extensions/validate_extensions.py:6]: also worth mentioning this is just a repurposing of the previous `generate_extension_db` script - hence the inclusion of the `/source` default etc

File: docs/requirements.txt
Comment (phlax) [reply docs/requirements.txt:38]: the extraneous changes here are produced by `pip-compile` not sure why they hadnt been set before - but im guessing that dependabot may be to blame for at least the hash ordering

File: tools/extensions/validate_extensions.py
Comment (phlax) [reply tools/extensions/validate_extensions.py:156]: ...and for that reason i think its important to keep it until the macro way has been removed completely

File: source/extensions/extensions_metadata.yaml
Comment (phlax) [reply source/extensions/extensions_metadata.yaml:574]: @htuch i have gone through all of the `BUILD` files with wasm in the path (and quite a few of the others) and none contain comments

can you point me to any specific examples please ?

**Comment by lizan** [comment]
I'm concerned about this as it is centralizing those metadata into a single file, which means it is harder for downstream builds to define those metadata in the extensions living there. While they won't included in our doc, but it is a good practice anyway. It is also making it harder when we move extensions to `contrib/`, separating some extensions from standard builds. Thoughts?

**Comment by phlax** [comment]
> Thoughts?

im very conscious of this - afaia the main downstream is envoy-mobile - so i have done quite a bit of work to figure out what will change and ease the transition

i dont think it will make it harder to separate contrib - but will defo require a slightly different path

File: source/extensions/extensions_metadata.yaml
Comment (htuch) [reply source/extensions/extensions_metadata.yaml:666]: Hah, not merged yet (I thought it had), see https://github.com/envoyproxy/envoy/pull/16057/files#diff-85815a26c8f765868f9c242a1968320d41e9c5798cbe94358339f3d99d0a70c2R16. We will need to update that PR.

File: tools/extensions/validate_extensions.py
Comment (htuch) [reply tools/extensions/validate_extensions.py:6]: Why do we need to merge this specific part of the PR if it goes in the next one?

File: tools/extensions/validate_extensions.py
Comment (htuch) [reply tools/extensions/validate_extensions.py:156]: If you want to preserve this over two PRs, then can you mark everything with a comment that will be gone in the next PR, so we don't have to review it with scrutiny?

File: tools/extensions/validate_extensions.py
Comment (phlax) [reply tools/extensions/validate_extensions.py:6]: its just maintaining the equivalence of `generate_extension_db` - ie ensuring it works in the same circumstances (of being called from `docs/build.sh`

File: tools/extensions/validate_extensions.py
Comment (phlax) [reply tools/extensions/validate_extensions.py:156]: ive added TODOs to all of the soon-to-be-removed code

**Review by htuch — APPROVED** [review]
Let's merge this as is and iterate on next step. Generally prefer to see everything complete before merging, but this will make the larger lift presumably easier with a checkpoint.

**Pull request merged**
